src/typ.rs

```diff
@@ -1341,6 +1529,43 @@ pub fn infer_module(
                 })
             }
 
+            Statement::TypeAlias {
+                meta,
+                public,
+                alias,
+                args,
+                resolved_type,
+            } => {
+                let mut type_vars = hashmap![];
+                let resolved_type_var_names = resolved_type.get_type_vars();
+
+                match env.type_from_ast(&resolved_type, &mut type_vars, NewTypeAction::MakeGeneric)
+                {
+                    Ok(typ) => {
+                        let generics = typ.bind_generic_args(&resolved_type_var_names);
```

Comment:
I think this approach could be made simpler by re-using existing functionality in the typer and by storing the type AST in the alias rather than the resolved type.

## When a type alias is defined:

1. Resolve the alias once to ensure the type is valid, then discard the resulting type
2. Store the alias name, the parameter names, and the AST of the aliased type

## When a type alias being resolved

1. Iterate across the alias args and insert the given parameter into `type_vars` for that name to seed the scope with those params.
2. Call `env.type_from_ast`, specifying `NewTypeAction::Disallow` to ensure there are no extra type variables that were not already declared.
3. Return the newly constructed type as the resolved type